### PR TITLE
feat: use flexible checkIndata + external checkInJudge, bytes type signature

### PR DESCRIPTION
**Pull request by ottbunn**

### CODE REVIEW

File: contracts/src/Tracker.sol [hunk 1]
```diff
@@ -97,17 +110,37 @@ contract Tracker is EIP712 {
         emit Join(msg.sender, challengeId);
     }
 
-    // todo: change signature to bytes, so we can support contract checkins
-    // todo: change checkin digest into bytes, to be more flexible
-    function checkIn(uint256 challengeId, uint256 timestamp, uint8 v, bytes32 r, bytes32 s) public {
+    /**
+     * challengeId is required to prevent replays from other challenges in case the same verifiers are used.
```
Comment (antoncoding) [inline contracts/src/Tracker.sol:114]: use natspec: 
@param challengeId
@param checkInData

File: contracts/src/Tracker.sol [hunk 2]
```diff
@@ -97,17 +110,37 @@ contract Tracker is EIP712 {
...
+     * 1. If the signing address matches the validator address
+     * 2. The challenge ID is the ID included in the signing digest.
+     * 3. The msg.sender is the user address included in the signing digest.
+     */
+    function checkIn(uint256 challengeId, bytes memory checkInData, bytes memory signature) public {
+        uint256 timestamp = block.timestamp;
         require(
-            timestamp <= challenges[challengeId].endTimestamp && timestamp >= challenges[challengeId].startTimestamp,
-            "invalid timestamp"
+            timestamp >= challenges[challengeId].startTimestamp && timestamp <= challenges[challengeId].endTimestamp,
```
Comment (antoncoding) [inline contracts/src/Tracker.sol:127]: nice change!

File: contracts/src/Tracker.sol [hunk 3]
```diff
@@ -97,17 +110,37 @@ contract Tracker is EIP712 {
...
     }
 
-    // todo: change signature to bytes, so we can support contract checkins
-    // todo: change checkin digest into bytes, to be more flexible
-    function checkIn(uint256 challengeId, uint256 timestamp, uint8 v, bytes32 r, bytes32 s) public {
+    /**
+     * challengeId is required to prevent replays from other challenges in case the same verifiers are used.
+     * checkInData is application specific, what to check on and replay prevention depend on the implementation of the checkInJudge contract specified at challenge creation.
+     * If the checkInJudge contract is specified, checkInData would be pass with the call checkInJudge.judge(checkInData).
+     * signature is generated with signing the returned digest from getCheckInDigest(uint256 challengeId, address user, bytes checkInData);
```
Comment (antoncoding) [inline contracts/src/Tracker.sol:117]: the comment here is a bit confusing, kind of implying the contract is generating the signature.
```suggestion
     * @signature checkInDigest signed by verifier address;
```

File: contracts/src/Tracker.sol [hunk 4]
```diff
@@ -97,17 +110,37 @@ contract Tracker is EIP712 {
         emit Join(msg.sender, challengeId);
     }
 
-    // todo: change signature to bytes, so we can support contract checkins
-    // todo: change checkin digest into bytes, to be more flexible
-    function checkIn(uint256 challengeId, uint256 timestamp, uint8 v, bytes32 r, bytes32 s) public {
+    /**
+     * challengeId is required to prevent replays from other challenges in case the same verifiers are used.
+     * checkInData is application specific, what to check on and replay prevention depend on the implementation of the checkInJudge contract specified at challenge creation.
+     * If the checkInJudge contract is specified, checkInData would be pass with the call checkInJudge.judge(checkInData).
```
Comment (antoncoding) [inline contracts/src/Tracker.sol:116]: move this to natspec @dev

**Review by antoncoding — APPROVED** [review]
Not bad!

**Review by ryanycw — APPROVED** [review]

**Pull request merged**